### ISSUE TITLE
Format Date / Calendar in toString for particular fields

### ISSUE THREAD

**Issue by jkaipa** [rzwitserloot/lombok#1741]
Looking for an annotation that would enable clients to specify expected format while toString is called on Date or Calendar or GregorianCalendar members. Any thoughts?

**Comment by Maaartinus** [comment]
@jkaipa There's no such thing, but you can use
```
@ToString.Exclude
private Calendar calendar;

@ToString.Include
private Calender calendar() {
    return MyFancyFormatter.format(calendar);
}
```

This is more flexible, but also more verbose than a general formatting applicable to all fields of type `Calendar`.

---

I could imagine something like
```
@ToString(transform=MyFancyFormatter::format)
```
but method references are not allowed in annotations. There could be
```
@ToString(transform=MyFancyFormatter.class)
```
which would require a class implementing `Function<?, ?>`. It'd require Java 8 and this may be a blocker as AFAIK Lombok so far works with Java 8.

Would it be better for you?

**Comment by axiopisty** [comment]
@Maaartinus I agree with @jkaipa. I think it would be useful to provide a generic annotation that could be applied to any field. See `@JsonFormat` as an example annotation from the jackson library.

https://github.com/FasterXML/jackson-annotations/blob/master/src/main/java/com/fasterxml/jackson/annotation/JsonFormat.java

**Comment by Maaartinus** [comment]
@axiopisty What Jackson does is interesting, but AFICT not directly applicable here. Note also that Jackson uses reflection for its magic, which is orders of magnitude simpler than what Lombok has to do. I guess, my

     @ToString(transform=MyFancyFormatter.class)

would cover most needs (maybe even all of them for the price of writing a few formatter classes). Together with a possibility to [define own aggregate annotations](https://github.com/rzwitserloot/lombok/issues/602) (another long wanted feature), it'd be pretty flexible.

However, I'm not a member of the team and I currently don't have the time to work on it. So you need an approval from a Lombok team member and a voluteer.

---

Note also that

    @ToString.Include private Whatever whatever() {return MyFancyFormatter.format(whatever);}

is sufficient (no need for `String.Exclude` as this happens automatically because of the method name being equal to the field name) for everything (depending on what you implemented in the formatter class).

**Comment by RezaGholamzad** [comment]
hello.
I agree with @axiopisty .
Are you @rzwitserloot going to generate this annotation?
Is there a problem adding this annotation?
If you intend, we will do this and add it to the project.
